File: plugins/buddy-pet/scripts/feeder.py
#!/usr/bin/env python3
"""
Feeder hook script for buddy-pet plugin.
Called by Claude Code hooks with mode: start, heartbeat, flush.

Reads JSONL transcript, extracts assistant events, sends feeding batches.
"""

import fcntl
import json
import math
import os
import sys
import time
import urllib.request

# Add scripts dir to path for common import
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import common

HEARTBEAT_INTERVAL = 30  # seconds (30 for testing, 300 for prod)

LOG_FILE = os.path.join(common.PLUGIN_DATA, "feeder.log")
LOKI_URL = os.environ.get("LOKI_URL", "http://localhost:3100")


def _log(msg):
    """Log to local file + push to Loki (best-effort)."""
    ts = time.strftime('%H:%M:%S')
    # Always write to local file
    try:
        os.makedirs(common.PLUGIN_DATA, exist_ok=True)
        with open(LOG_FILE, "a") as f:
            f.write(f"{ts} {msg}\n")
    except OSError:
        pass
    # Push to Loki (fire-and-forget, never blocks feeder)
    try:
        ns = str(int(time.time() * 1e9))
        payload = json.dumps({
            "streams": [{
                "stream": {"service": "feeder", "job": "buddy-pet-plugin"},
                "values": [[ns, f"{ts} {msg}"]]
            }]
        }).encode("utf-8")
        req = urllib.request.Request(
            LOKI_URL + "/loki/api/v1/push",
            data=payload,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        urllib.request.urlopen(req, timeout=2)
    except Exception:
        pass  # Never fail the feeder because of logging


# ---------------------------------------------------------------------------
# Auto-birth: ensure buddy exists for current account
# ---------------------------------------------------------------------------


def _ensure_config():
    """Load config for current account. Auto-birth if missing. Returns config or None."""
    config = common.load_config()
    if config and config.get("buddy_token"):
        return config
    return _auto_birth()


def _auto_birth():
    """Auto-register buddy for the current account. Returns config or None."""
    user_id = common.get_user_id()
    if user_id == "anon":
        _log("auto-birth: userId=anon, skip")
        return None

    _log("auto-birth: registering buddy...")

    user_hash = common.compute_user_hash(user_id)
    instance_id = common.compute_instance_id()

    # Read companion (may be None if user hasn't hatched yet)
    companion = common.read_companion_data()
    birth_data = {
        "user_hash": user_hash,
        "instance_id": instance_id,
        "plugin_version": common.PLUGIN_VERSION,
    }
    if companion:
        birth_data["companion"] = {
            "name": companion.get("name", ""),
            "personality": companion.get("personality", ""),
            "hatched_at": companion.get("hatched_at", 0),
            "species": companion.get("species", ""),
            "rarity": companion.get("rarity", ""),
            "eye": companion.get("eye", ""),
            "hat": companion.get("hat", ""),
            "shiny": companion.get("shiny", False),
            "base_stats": companion.get("base_stats", {}),
        }
    else:
        # Minimal companion — backend will accept it
        birth_data["companion"] = {
            "name": "Buddy", "personality": "", "species": "blob",
            "rarity": "common", "eye": "o", "hat": "none", "shiny": False,
            "hatched_at": 0, "base_stats": {},
        }

    status, resp = common.http_post("/guild/buddy/birth", birth_data)

    if status not in (200, 201):
        _log(f"auto-birth: failed HTTP {status}: {resp}")
        return None

    token = resp.get("buddy_token", "")
    buddy_id = resp.get("buddy_id", "")
    if not token:
        _log(f"auto-birth: no token in response: {resp}")
        return None

    config = {
        "buddy_token": token,
        "buddy_id": buddy_id,
        "api_url": common.get_api_url(),
    }
    common.save_config(config)

    name = resp.get("buddy", {}).get("name", buddy_id[:8])
    _log(f"auto-birth: OK — {name} ({buddy_id})")
    return config


def main():
    if len(sys.argv) < 2:
        sys.exit(0)

    mode = sys.argv[1]
    _log(f"--- {mode} ---")

    if mode == "start":
        do_start()
    elif mode == "heartbeat":
        do_heartbeat()
    elif mode == "flush":
        do_flush()


# ---------------------------------------------------------------------------
# Mode: start
# ---------------------------------------------------------------------------


def do_start():
    """SessionStart: auto-birth if needed, retry pending reports."""
    config = _ensure_config()
    if not config:
        return

    # Read stdin hook input (contains session_id, transcript_path)
    _read_hook_input()

    # Submit pending reports
    submit_pending(config["buddy_token"])


# ---------------------------------------------------------------------------
# Mode: heartbeat
# ---------------------------------------------------------------------------


def do_heartbeat():
    """UserPromptSubmit: send feeding if interval since last send."""
    config = _ensure_config()
    if not config:
        return

    hook_input = _read_hook_input()
    if not hook_input:
        _log("heartbeat: no hook_input, skip")
        return

    state = common.load_state()

    # Check heartbeat interval
    now = time.time()
    elapsed = now - state.get("last_send_time", 0)
    if elapsed < HEARTBEAT_INTERVAL:
        _log(f"heartbeat: {elapsed:.0f}s < {HEARTBEAT_INTERVAL}s, skip")
        return

    _log(f"heartbeat: {elapsed:.0f}s elapsed, sending...")
    _send_events(config, hook_input, state)


# ---------------------------------------------------------------------------
# Mode: flush
# ---------------------------------------------------------------------------


def do_flush():
    """SessionEnd: send all remaining events (no time check)."""
    config = _ensure_config()
    if not config:
        return

    hook_input = _read_hook_input()
    if not hook_input:
        _log("flush: no hook_input, skip")
        return

    _log("flush: sending all remaining events")

    state = common.load_state()
    _send_events(config, hook_input, state)


# ---------------------------------------------------------------------------
# Core: read JSONL, extract events, send
# ---------------------------------------------------------------------------


def _send_events(config, hook_input, state):
    """Acquire lock, read JSONL, extract events, POST /feeding."""
    session_id = hook_input.get("session_id", "")
    transcript_path = hook_input.get("transcript_path", "")
    cwd = hook_input.get("cwd", "")

    if not transcript_path or not os.path.isfile(transcript_path):
        _log(f"send: no transcript at {transcript_path}")
        return

    # Acquire file lock (non-blocking)
    lock_fd = _acquire_lock()
    if lock_fd is None:
        _log("send: lock held by another process")
        return

    try:
        # Reload state under lock (may have changed)
        state = common.load_state()

        # Get session state
        sessions = state.get("sessions", {})
        session_state = sessions.get(session_id, {"offset": 0, "batch_seq": 0})
        offset = session_state.get("offset", 0)

        # Read JSONL from offset
        events, new_offset = _read_jsonl_events(transcript_path, offset)

        if not events:
            _log(f"send: no new events (offset={offset})")
            return

        # Check companion sync
        companion_sync = None
        current_companion_hash = common.companion_hash_string()
        if current_companion_hash and current_companion_hash != state.get("companion_hash", ""):
            companion_data = common.read_companion_data()
            if companion_data:
                companion_sync = companion_data
                _log("send: companion sync included (hash changed)")

        # Build payload
        batch_seq = int(time.time())
        payload = {
            "session_hash": common.compute_session_hash(session_id),
            "project_hash": common.compute_project_hash(cwd),
            "batch_seq": batch_seq,
            "plugin_version": common.PLUGIN_VERSION,
            "companion_sync": companion_sync,
            "events": events,
        }

        # POST /feeding
        _log(f"send: POST /feeding {len(events)} events, offset {offset}->{new_offset}")
        status, resp = common.http_post("/feeding", payload, token=config["buddy_token"])
        _log(f"send: HTTP {status} → {resp}")

        if status == 401:
            # Token invalid — re-register and retry once
            _log("send: 401 — token invalid, re-registering")
            common.delete_config()
            new_config = _auto_birth()
            if new_config:
                payload["session_hash"] = common.compute_session_hash(session_id)
                status, resp = common.http_post("/feeding", payload, token=new_config["buddy_token"])
                _log(f"send: retry HTTP {status} → {resp}")

        if status in (202, 409):
            # Success or duplicate — update state
            session_state["offset"] = new_offset
            session_state["batch_seq"] = batch_seq
            sessions[session_id] = session_state
            state["sessions"] = sessions
            state["last_send_time"] = time.time()
            if current_companion_hash:
                state["companion_hash"] = current_companion_hash
            common.save_state(state)
        elif status != 401:
            # Non-auth failure — save to pending for retry
            _log(f"send: HTTP {status} non-success, saving to pending")
            _save_pending(payload)

    finally:
        _release_lock(lock_fd)


def _read_jsonl_events(transcript_path, offset):
    """
    Read JSONL file from byte offset, extract assistant and compact_boundary events.
    Returns (events_list, new_byte_offset).
    """
    events = []

    try:
        with open(transcript_path, "rb") as f:
            f.seek(offset)
            raw = f.read()
            new_offset = offset + len(raw)
    except (IOError, OSError):
        return [], offset

    for line in raw.split(b"\n"):
        if not line.strip():
            continue
        try:
            entry = json.loads(line)
        except (json.JSONDecodeError, UnicodeDecodeError):
            continue

        entry_type = entry.get("type")

        # Assistant messages with model response
        if entry_type == "assistant" and "message" in entry:
            event = _extract_assistant_event(entry)
            if event:
                events.append(event)

        # Compact boundary
        elif entry_type == "system" and entry.get("subtype") == "compact_boundary":
            event = _extract_compact_event(entry)
            if event:
                events.append(event)

    return events, new_offset


def _extract_assistant_event(entry):
    """Extract event fields from an assistant JSONL entry."""
    msg = entry.get("message", {})
    usage = msg.get("usage", {})
    content = msg.get("content", [])
    server_tool_use = usage.get("server_tool_use", {})

    # Parse timestamp: ISO 8601 to unix millis
    e_ts = _iso_to_millis(entry.get("timestamp", ""))
    if e_ts is None:
        return None

    # Content analysis
    c_types = []
    c_tu_n = []
    c_tu_ids = []
    c_ts = False

    for block in content:
        block_type = block.get("type", "")
        if block_type and block_type not in c_types:
            c_types.append(block_type)

        if block_type == "thinking":
            if block.get("signature"):
                c_ts = True

        elif block_type == "tool_use":
            tool_name = block.get("name", "")
            tool_id = block.get("id", "")
            if tool_name and tool_name not in c_tu_n:
                c_tu_n.append(tool_name)
            if tool_id:
                c_tu_ids.append(tool_id)

    # Build event with exact gateway field names
    event = {
        "e_ts": str(e_ts),
        "e_t": "assistant",
        "m_id": msg.get("id", ""),
        "r_id": entry.get("requestId", ""),
        "model": msg.get("model", ""),
        "u_ot": usage.get("output_tokens", 0),
        "u_it": usage.get("input_tokens", 0),
        "u_crt": usage.get("cache_read_input_tokens", 0),
        "u_cwt": usage.get("cache_creation_input_tokens", 0),
        "u_ws": server_tool_use.get("web_search_requests", 0),
        "c_types": c_types,
        "c_tu_n": c_tu_n,
        "c_tu_ids": c_tu_ids,
        "c_ts": c_ts,
        "stop": msg.get("stop_reason"),  # null, "end_turn", "tool_use" — keep as-is
    }

    return event


def _extract_compact_event(entry):
    """Extract event fields from a compact_boundary JSONL entry."""
    e_ts = _iso_to_millis(entry.get("timestamp", ""))
    if e_ts is None:
        return None

    metadata = entry.get("compactMetadata", {})

    event = {
        "e_ts": str(e_ts),
        "e_t": "compact",
        "compact_pre_tokens": metadata.get("preTokens", 0),
    }

    return event


def _iso_to_millis(iso_str):
    """Parse ISO 8601 timestamp to unix milliseconds."""
    if not iso_str:
        return None
    try:
        # Handle "2026-04-01T08:57:54.470Z" format
        # Strip trailing Z and parse
        s = iso_str.rstrip("Z")
        # Split date and time
        if "T" not in s:
            return None
        date_part, time_part = s.split("T", 1)
        year, month, day = date_part.split("-")

        # Handle fractional seconds
        if "." in time_part:
            time_main, frac = time_part.split(".", 1)
            # Pad or truncate frac to 3 digits (millis)
            frac = frac[:3].ljust(3, "0")
            millis_frac = int(frac)
        else:
            time_main = time_part
            millis_frac = 0

        hour, minute, second = time_main.split(":")

        import calendar
        import datetime

        dt = datetime.datetime(
            int(year), int(month), int(day),
            int(hour), int(minute), int(second),
            tzinfo=datetime.timezone.utc,
        )
        epoch_seconds = int(calendar.timegm(dt.timetuple()))
        return epoch_seconds * 1000 + millis_frac

    except (ValueError, IndexError, AttributeError):
        return None


# ---------------------------------------------------------------------------
# Pending reports
# ---------------------------------------------------------------------------


def submit_pending(token):
    """Retry sending pending reports from PENDING_DIR."""
    if not os.path.isdir(common.PENDING_DIR):
        return

    try:
        files = sorted(os.listdir(common.PENDING_DIR))
    except OSError:
        return

    pending_json = [f for f in files if f.endswith(".json")]
    if pending_json:
        _log(f"pending: {len(pending_json)} files to retry")

    for fname in pending_json:
        fpath = os.path.join(common.PENDING_DIR, fname)
        try:
            with open(fpath, "r") as f:
                payload = json.load(f)
        except (IOError, json.JSONDecodeError):
            continue

        status, _resp = common.http_post("/feeding", payload, token=token)
        if status in (202, 409):
            _log(f"pending: sent {fname} (HTTP {status})")
            # Sent or duplicate — remove
            try:
                os.remove(fpath)
            except OSError:
                pass
        else:
            _log(f"pending: retry failed {fname} (HTTP {status})")


def _save_pending(payload):
    """Save a failed payload to PENDING_DIR for later retry."""
    os.makedirs(common.PENDING_DIR, exist_ok=True)
    fname = "{:.0f}.json".format(time.time() * 1000)
    fpath = os.path.join(common.PENDING_DIR, fname)
    try:
        with open(fpath, "w") as f:
            json.dump(payload, f)
    except (IOError, OSError):
        pass


# ---------------------------------------------------------------------------
# File locking
# ---------------------------------------------------------------------------


def _acquire_lock():
    """Acquire exclusive file lock (non-blocking). Returns fd or None."""
    os.makedirs(common.PLUGIN_DATA, exist_ok=True)
    try:
        fd = os.open(common.LOCK_FILE, os.O_CREAT | os.O_RDWR)
        fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
        return fd
    except (IOError, OSError):
        return None


def _release_lock(fd):
    """Release file lock and close fd."""
    if fd is not None:
        try:
            fcntl.flock(fd, fcntl.LOCK_UN)
            os.close(fd)
        except (IOError, OSError):
            pass


# ---------------------------------------------------------------------------
# Hook stdin
# ---------------------------------------------------------------------------


def _read_hook_input():
    """Read JSON from stdin (piped by Claude Code hook system)."""
    try:
        raw = sys.stdin.read()
        if raw:
            return json.loads(raw)
    except (json.JSONDecodeError, IOError):
        pass
    return {}


if __name__ == "__main__":
    main()

File: plugins/buddy-pet/scripts/feeder.js
#!/usr/bin/env node
/**
 * Feeder hook script for buddy-pet plugin.
 * Called on SessionStart, UserPromptSubmit, SessionEnd.
 */
'use strict';

const fs = require('fs');
const path = require('path');
const common = require('./common');

// Catch ALL errors — hooks must never show errors to user.
process.on('uncaughtException', () => process.exit(0));
process.on('unhandledRejection', () => process.exit(0));

const HEARTBEAT_INTERVAL = 30; // seconds
const CONSENT_STALE_SECONDS = 300; // 5 minutes

// ---------------------------------------------------------------------------
// Logging
// ---------------------------------------------------------------------------

const LOG_FILE = path.join(common.PLUGIN_DATA, 'feeder.log');
const LOG_MAX_BYTES = 512 * 1024; // 512 KB
const LOG_KEEP_BYTES = 256 * 1024; // keep last 256 KB after rotation

function rotateLogIfNeeded() {
  try {
    const stat = fs.statSync(LOG_FILE);
    if (stat.size <= LOG_MAX_BYTES) return;
    const buf = Buffer.alloc(LOG_KEEP_BYTES);
    const fd = fs.openSync(LOG_FILE, 'r');
    const bytesRead = fs.readSync(fd, buf, 0, LOG_KEEP_BYTES, stat.size - LOG_KEEP_BYTES);
    fs.closeSync(fd);
    const data = buf.subarray(0, bytesRead);
    // Find first newline to avoid partial line at start
    const nl = data.indexOf(10); // '\n'
    const clean = nl >= 0 ? data.subarray(nl + 1) : data;
    fs.writeFileSync(LOG_FILE, clean);
  } catch {}
}

function log(msg) {
  const ts = new Date().toISOString();
  const line = `[${ts}] ${msg}\n`;
  try { fs.appendFileSync(LOG_FILE, line); } catch {}
}

// ---------------------------------------------------------------------------
// Auto-birth
// ---------------------------------------------------------------------------

async function ensureConfig() {
  let config = common.loadConfig();
  if (config && config.buddy_token) return config;
  return autoBirth();
}

async function autoBirth() {
  const userId = common.getUserId();
  if (userId === 'anon') { log('auto-birth: userId=anon, skip'); return null; }
  const companion = common.readCompanionData() || {
    name: 'Buddy', personality: '', hatched_at: 0,
    species: 'blob', rarity: 'common', eye: '.', hat: 'none',
    shiny: false, base_stats: { DEBUGGING: 50, PATIENCE: 50, CHAOS: 50, WISDOM: 50, SNARK: 50 },
  };

  const birthData = {
    user_hash: common.computeUserHash(userId),
    instance_id: common.computeInstanceId(),
    companion: {
      name: companion.name, personality: companion.personality,
      hatched_at: companion.hatched_at, species: companion.species,
      rarity: companion.rarity, eye: companion.eye, hat: companion.hat,
      shiny: companion.shiny, base_stats: companion.base_stats,
    },
    plugin_version: common.PLUGIN_VERSION,
  };

  const [status, resp] = await common.httpPost('/guild/buddy/birth', birthData);

  if (status === 200 || status === 201) {
    const config = {
      buddy_token: resp.buddy_token,
      buddy_id: resp.buddy_id,
      api_url: common.getApiUrl(),
    };
    common.saveConfig(config);
    log(`auto-birth: buddy_id=${resp.buddy_id}`);
    return config;
  }

  log(`auto-birth failed: ${status} ${JSON.stringify(resp)}`);
  return null;
}

// ---------------------------------------------------------------------------
// Hook input
// ---------------------------------------------------------------------------

function readHookInput() {
  try {
    const raw = fs.readFileSync(0, 'utf8'); // stdin
    return JSON.parse(raw);
  } catch {
    return {};
  }
}

// ---------------------------------------------------------------------------
// JSONL reading
// ---------------------------------------------------------------------------

function readJsonlEvents(filePath, byteOffset) {
  if (!filePath || !fs.existsSync(filePath)) return { events: [], newOffset: byteOffset };

  const fd = fs.openSync(filePath, 'r');
  const stat = fs.fstatSync(fd);
  const toRead = stat.size - byteOffset;
  if (toRead <= 0) { fs.closeSync(fd); return { events: [], newOffset: byteOffset }; }

  const buf = Buffer.alloc(toRead);
  fs.readSync(fd, buf, 0, toRead, byteOffset);
  fs.closeSync(fd);

  const lines = buf.toString('utf8').split('\n').filter(Boolean);
  const events = [];

  for (const line of lines) {
    try {
      const obj = JSON.parse(line);
      if (obj.type === 'assistant' && obj.message) {
        const ev = extractAssistantEvent(obj);
        if (ev) events.push(ev);
      } else if (obj.type === 'system' && obj.subtype === 'compact_boundary') {
        const ev = extractCompactEvent(obj);
        if (ev) events.push(ev);
      }
    } catch {}
  }

  return { events, newOffset: byteOffset + toRead };
}

function isoToMillis(iso) {
  if (!iso) return null;
  try {
    const ms = new Date(iso).getTime();
    return isNaN(ms) ? null : String(ms);
  } catch { return null; }
}

function extractAssistantEvent(obj) {
  const ets = isoToMillis(obj.timestamp);
  if (ets === null) return null;

  const msg = obj.message || {};
  const usage = msg.usage || {};
  const content = msg.content || [];

  const cTypes = [...new Set(content.map(b => b.type).filter(Boolean))];
  const toolUses = content.filter(b => b.type === 'tool_use');
  const cTuN = [...new Set(toolUses.map(b => b.name).filter(Boolean))];
  const cTuIds = toolUses.map(b => b.id).filter(Boolean);
  const hasThinkingSig = content.some(b => b.type === 'thinking' && b.signature);

  // web_search from server_tool_use
  let uWs = 0;
  if (usage.server_tool_use) {
    uWs = usage.server_tool_use.web_search_requests || 0;
  }

  return {
    e_ts: ets,
    e_t: 'assistant',
    m_id: msg.id || '',
    r_id: obj.requestId || '',
    model: msg.model || '',
    u_ot: usage.output_tokens || 0,
    u_it: usage.input_tokens || 0,
    u_crt: usage.cache_read_input_tokens || 0,
    u_cwt: usage.cache_creation_input_tokens || 0,
    u_ws: uWs,
    c_types: cTypes,
    c_tu_n: cTuN,
    c_tu_ids: cTuIds,
    c_ts: hasThinkingSig,
    stop: msg.stop_reason || null,
  };
}

function extractCompactEvent(obj) {
  const ets = isoToMillis(obj.timestamp);
  if (ets === null) return null;
  const meta = obj.compactMetadata || {};
  return { e_ts: ets, e_t: 'compact', compact_pre_tokens: meta.preTokens || 0 };
}

// ---------------------------------------------------------------------------
// Pending system
// ---------------------------------------------------------------------------

async function submitPending(token) {
  common.ensureDir(common.PENDING_DIR);
  let files;
  try { files = fs.readdirSync(common.PENDING_DIR).filter(f => f.endsWith('.json')); } catch { return; }

  await Promise.all(files.map(async (file) => {
    const fp = path.join(common.PENDING_DIR, file);
    try {
      const payload = JSON.parse(fs.readFileSync(fp, 'utf8'));
      const [status] = await common.httpPost('/feeding', payload, token);
      if (status === 202 || status === 409) {
        try { fs.unlinkSync(fp); } catch {}
      }
    } catch {}
  }));
}

function savePending(payload) {
  common.ensureDir(common.PENDING_DIR);
  const fp = path.join(common.PENDING_DIR, `${Date.now()}.json`);
  fs.writeFileSync(fp, JSON.stringify(payload));
}

// ---------------------------------------------------------------------------
// Consent: fetch, cache, filter
// ---------------------------------------------------------------------------

/**
 * Load consents from local config cache.
 * Returns object like { token_usage: true, tool_usage: false, ... } or null.
 */
function loadCachedConsents(config) {
  if (!config.consents || !config.consents_fetched_at) return null;
  return config.consents;
}

/**
 * Check if local consent cache is stale (older than 5 minutes).
 */
function isConsentStale(config) {
  if (!config.consents_fetched_at) return true;
  const age = Math.floor(Date.now() / 1000) - config.consents_fetched_at;
  return age > CONSENT_STALE_SECONDS;
}

/**
 * Fetch consents from API, update local cache.
 * Fail-open: returns cached or all-enabled on error.
 */
async function refreshConsents(config) {
  try {
    const [status, resp] = await common.httpGet('/buddy/me/consents', config.buddy_token);
    if (status === 200 && resp.consents) {
      const consents = {};
      for (const c of resp.consents) {
        consents[c.key] = c.enabled;
      }
      config.consents = consents;
      config.consents_fetched_at = Math.floor(Date.now() / 1000);
      common.saveConfig(config);
      log('consents refreshed');
      return consents;
    }
    log(`consent fetch non-200: ${status}`);
  } catch (err) {
    log(`consent fetch error: ${err.message}`);
  }
  // Fail-open: use cached or assume all-enabled
  return loadCachedConsents(config) || allEnabled();
}

function allEnabled() {
  return {
    token_usage: true,
    tool_usage: true,
    model_identity: true,
    thinking_mode: true,
    web_search: true,
    session_timing: true,
    project_identity: true,
  };
}

/**
 * Get consents for filtering: use cache if fresh, otherwise refresh.
 */
async function getConsents(config) {
  if (!isConsentStale(config)) {
    return loadCachedConsents(config) || allEnabled();
  }
  return refreshConsents(config);
}

/**
 * Apply consent filter to a single event (mutates in place).
 */
function applyConsentFilter(event, consents) {
  if (!consents) return;

  // token_usage disabled: zero token fields
  if (consents.token_usage === false) {
    event.u_ot = 0;
    event.u_it = 0;
    event.u_crt = 0;
    event.u_cwt = 0;
    if ('compact_pre_tokens' in event) event.compact_pre_tokens = 0;
  }

  // tool_usage disabled: remove tool_use from c_types, zero tool arrays
  if (consents.tool_usage === false) {
    if (Array.isArray(event.c_types)) {
      event.c_types = event.c_types.filter(t => t !== 'tool_use');
    }
    event.c_tu_n = [];
    event.c_tu_ids = [];
  }

  // model_identity disabled: blank model
  if (consents.model_identity === false) {
    event.model = '';
  }

  // thinking_mode disabled: remove thinking from c_types, clear thinking sig
  if (consents.thinking_mode === false) {
    if (Array.isArray(event.c_types)) {
      event.c_types = event.c_types.filter(t => t !== 'thinking');
    }
    event.c_ts = false;
  }

  // web_search disabled: zero web search count
  if (consents.web_search === false) {
    event.u_ws = 0;
  }
}

// ---------------------------------------------------------------------------
// Core send
// ---------------------------------------------------------------------------

async function sendEvents(config, hookInput, state) {
  const sessionId = hookInput.session_id || '';
  const transcriptPath = hookInput.transcript_path || '';
  const cwd = hookInput.cwd || process.cwd();

  if (!transcriptPath || !sessionId) return;

  if (!common.acquireLock()) { log('lock held, skipping'); return; }

  try {
    // Reload state under lock
    state = common.loadState();

    const sessionKey = common.computeSessionHash(sessionId);
    const sessionState = state.sessions[sessionKey] || { offset: 0, batch_seq: 0 };

    const { events, newOffset } = readJsonlEvents(transcriptPath, sessionState.offset);

    if (events.length === 0 && !companionChanged(state)) {
      sessionState.offset = newOffset;
      state.sessions[sessionKey] = sessionState;
      state.last_send_time = Date.now() / 1000;
      common.saveState(state);
      return;
    }

    // Client-side consent filtering (fail-open)
    const consents = await getConsents(config);
    for (const ev of events) {
      applyConsentFilter(ev, consents);
    }

    const projectHash = consents.project_identity === false
      ? '' : common.computeProjectHash(cwd);

    const payload = {
      session_hash: sessionKey,
      project_hash: projectHash,
      batch_seq: Math.floor(Date.now() / 1000),
      plugin_version: common.PLUGIN_VERSION,
      events,
    };

    // Companion sync if changed
    const newCompHash = common.companionHashString();
    if (newCompHash && newCompHash !== state.companion_hash) {
      const comp = common.readCompanionData();
      if (comp) payload.companion_sync = comp;
    }

    const [status, resp] = await common.httpPost('/feeding', payload, config.buddy_token);

    if (status === 401) {
      log('401 — re-registering');
      common.deleteConfig();
      const newConfig = await autoBirth();
      if (newConfig) {
        const [s2] = await common.httpPost('/feeding', payload, newConfig.buddy_token);
        if (s2 === 202 || s2 === 409) {
          updateSessionState(state, sessionKey, newOffset, newCompHash);
        }
      }
      return;
    }

    if (status === 202 || status === 409) {
      updateSessionState(state, sessionKey, newOffset, newCompHash);
      log(`fed: ${events.length} events, status=${status}`);
    } else {
      log(`feed failed: ${status} ${JSON.stringify(resp)}`);
      savePending(payload);
    }
  } finally {
    common.releaseLock();
  }
}

function companionChanged(state) {
  const h = common.companionHashString();
  return h && h !== state.companion_hash;
}

function updateSessionState(state, sessionKey, newOffset, newCompHash) {
  const ss = state.sessions[sessionKey] || { offset: 0, batch_seq: 0 };
  ss.offset = newOffset;
  ss.batch_seq = Math.floor(Date.now() / 1000);
  state.sessions[sessionKey] = ss;
  state.last_send_time = Date.now() / 1000;
  if (newCompHash) state.companion_hash = newCompHash;

  // Prune sessions older than 7 days
  const cutoff = Date.now() / 1000 - 7 * 86400;
  for (const key of Object.keys(state.sessions)) {
    if ((state.sessions[key].batch_seq || 0) < cutoff) delete state.sessions[key];
  }

  common.saveState(state);
}

// ---------------------------------------------------------------------------
// Modes
// ---------------------------------------------------------------------------

async function doStart(hookInput) {
  const config = await ensureConfig();
  if (!config) return;
  // Refresh consents on session start
  await refreshConsents(config);
  await submitPending(config.buddy_token);
}

async function doHeartbeat(hookInput) {
  const config = await ensureConfig();
  if (!config) return;

  const state = common.loadState();
  const elapsed = Date.now() / 1000 - state.last_send_time;
  if (elapsed < HEARTBEAT_INTERVAL) return;

  await sendEvents(config, hookInput, state);
}

async function doFlush(hookInput) {
  const config = await ensureConfig();
  if (!config) return;

  const state = common.loadState();
  await sendEvents(config, hookInput, state);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const mode = process.argv[2];
  if (!mode) return;

  rotateLogIfNeeded();
  const hookInput = readHookInput();

  switch (mode) {
    case 'start': await doStart(hookInput); break;
    case 'heartbeat': await doHeartbeat(hookInput); break;
    case 'flush': await doFlush(hookInput); break;
  }
}

main().catch((err) => {
  try { log(`error: ${err.message}`); } catch {}
  // Never exit with non-zero — hook errors are shown to user and it's not their fault.
  process.exit(0);
});
